src/ProtocolFeeController.sol

```diff
@@ -0,0 +1,117 @@
+
+        defaultProtocolFees[key.toId()] = newProtocolFee;
+    }
+
+    /// @notice Get the protocol fee for a pool given the conditions of this contract
+    /// @return protocolFee The pool's protocol fee, expressed in hundredths of a bip. The upper 12 bits are for 1->0
+    function protocolFeeForPool(PoolKey memory poolKey) external view override returns (uint24 protocolFee) {
+        if (address(poolKey.poolManager) != poolManager) revert InvalidPoolManager();
+
+        // in case we've set a default protocol fee for the pool
+        uint24 defaultProtocolFee = defaultProtocolFees[poolKey.toId()];
+        if (defaultProtocolFee != 0) {
+            // already bi-directional fee so that it can be different in each direction
+            return defaultProtocolFee;
+        }
+
+        // otherwise, calculate the protocol fee based on the predefined rule
+        uint256 lpFee = poolKey.fee;
+        if (lpFee == LPFeeLibrary.DYNAMIC_FEE_FLAG) {
+            /// @notice for dynamic fee pools, the default protocol fee is 0
+            return _buildProtocolFee(0);
+        } else {
+            /// @notice for static fee pools, the protocol fee should be a portion of the total fee based on 'protocolFeeSplitRatio'
+            /// @dev the formula is derived from the following equation:
+            /// totalSwapFee = protocolFee + (1 - protocolFee) * lpFee = protocolFee / protocolFeeSplitRatio
+            uint24 oneDirectionProtocolFee = uint24(
+                lpFee * ONE_HUNDRED_PERCENT_RATIO
+                    / (
+                        lpFee + ONE_HUNDRED_PERCENT_RATIO * ONE_HUNDRED_PERCENT_RATIO / protocolFeeSplitRatio
+                            - ONE_HUNDRED_PERCENT_RATIO
+                    )
+            );
+
+            // cap the protocol fee at 0.4%, if it's over the limit we set it to the max
+            if (oneDirectionProtocolFee > ProtocolFeeLibrary.MAX_PROTOCOL_FEE) {
+                oneDirectionProtocolFee = ProtocolFeeLibrary.MAX_PROTOCOL_FEE;
+            }
+            return _buildProtocolFee(oneDirectionProtocolFee);
+        }
+    }
+
+    /// @param fee If 1000, the protocol fee is 0.1%, cap at 0.4%
+    function _buildProtocolFee(uint24 fee) internal pure returns (uint24) {
+        return fee + (fee << 12);
+    }
+
+    /// @notice Override the default protcool fee for the pool
+    /// @dev this could be used for marketing campaign where PCS takes 0 protocol fee for a pool for a period
+    function setProtocolFee(PoolKey memory key, uint24 newProtocolFee) external onlyOwner {
+        if (address(key.poolManager) != poolManager) revert InvalidPoolManager();
+
+        // no need to validate the protocol fee as it will be done in the pool manager
+        IProtocolFees(address(key.poolManager)).setProtocolFee(key, newProtocolFee);
+    }
+
+    function collectProtocolFee(address recipient, Currency currency, uint256 amount) external onlyOwner {
```

Comment:
i notice there is no `collectProtocolFee` in `CLPoolManagerOwner`,  i'd like to confirm if ProtocolFeeController's owner will be CLPoolManagerOwner ?

if yes, we need to add collectProtocolFee function in CLPoolManagerOwner as well otherwise no one can collect protocol fee

Reply:
i prefer to use different owner , keep PoolManagerOwner transactions clear

Reply:
> i notice there is no collectProtocolFee in CLPoolManagerOwner, i'd like to confirm if ProtocolFeeController's owner will be CLPoolManagerOwner ?

only protocolFeeController can collect protocol fee right?

Reply:
"Only protocolFeeController can collect protocol fee right? owner of poolManager cannot call poolManager.collectProtocolFee"

Yes, it used to support both but we've updated the logic here

Reply:
can we mark this as resolved too? since we decided to go with just 1 owner -- and the owner can call both `collectProtocolFee` and `setProtocolFee(`